core/type_inference/translate_jsonnet_to_lambda.py

```diff
@@ -0,0 +1,192 @@
+
+    elif isinstance(ast_, j_ast.BuiltinFunction):
+        raise Exception('Not translated yet!\n')
+
+    elif isinstance(ast_, j_ast.Conditional):
+        raise Exception('Not translated yet!\n')
+
+    elif isinstance(ast_, j_ast.Error):
+        raise Exception('Not translated yet!\n')
+
+    elif isinstance(ast_, j_ast.Function):
+        def build_lambda(args, body):
+            if not args:
+                return translate_to_lambda_ast(body, my_env)
+            return lam_ast.Lambda(args[0].id, build_lambda(args[1:], body))
+        return build_lambda(ast_.arguments, ast_.body)
+
+    elif isinstance(ast_, j_ast.InSuper):
+        raise Exception('Not translated yet!\n')
+
+    elif isinstance(ast_, j_ast.Index):
+        location = translate_location(ast_.location)
+        if isinstance(ast_.target, j_ast.Self) and isinstance(ast_.index, j_ast.LiteralString):
+            return lam_ast.Identifier(ast_.index.value, location)
+        else:
+            raise Exception('Not translated yet!\n')
+
+    elif isinstance(ast_, j_ast.LiteralBoolean):
+        location = translate_location(ast_.location)
+        return lam_ast.LiteralBoolean(ast_.value, location)
+
+    elif isinstance(ast_, j_ast.LiteralNumber):
+        location = translate_location(ast_.location)
+        return lam_ast.LiteralNumber(ast_.value, location)
+
+    elif isinstance(ast_, j_ast.LiteralString):
+        location = translate_location(ast_.location)
+        return lam_ast.LiteralString(ast_.value, location)
+
+    elif isinstance(ast_, j_ast.LiteralNull):
+        return lam_ast.Identifier("null")
+
+    elif isinstance(ast_, j_ast.ObjectComprehension):
+        raise Exception('Not translated yet!\n')
+
+    elif isinstance(ast_, j_ast.Self):
+        return lam_ast.Identifier("self")
+
+    elif isinstance(ast_, j_ast.SuperIndex):
+        return lam_ast.Identifier(ast_.index.value)
+
+    elif isinstance(ast_, j_ast.UnaryOp):
+        raise Exception('Not translated yet!\n')
+
+    elif isinstance(ast_, j_ast.Var):
+        location = translate_location(ast_.location)
+        return lam_ast.Identifier(ast_.id, location)
+
+    else:
+        print(ast_.__class__)
+        raise Exception('Node is not found in jsonnet_ast\n')
+
+
+def build_record_type_constructor(fields):
+    var_type = {}
+    name_type = {}
+    for i, field in enumerate(fields):
+        name, _ = field
+        var = f'var{i}'
+        var_type[var] = TypeVariable()
+        name_type[name] = var_type[var]
+    record_type = TypeRowOperator(name_type)
+
+    def rec_build(i, n, var_type, record_type):
+        if n == 0:
+            return record_type
+        var = var_type[f'var{i}']
+        if (i == n-1):
+            return Function(var, record_type)
+        return Function(var, rec_build(i+1, n, var_type, record_type))
+
+    return rec_build(0, len(var_type), var_type, record_type)
+
+
+def build_letrec_and(field_keys, fields, body, env, location):
```

Comment:
Can you add some comments to all of these functions explaining roughly what they do. Maybe you can copy-paste some stuff out of the design doc as a start.

Reply:
ok